ceph/rados/core_workflows.py

```diff
@@ -2386,32 +2387,146 @@ def add_network_delay_on_host(
+        Args:
+            action: operation to be performed on the service, i.e. start, stop, restart, disable, enable
+            daemon_type: ceph service type (mon, mgr ...)
+            daemon_id: Name of the service
+            timeout: time for which the method would wait for state change
+
+        Note: A separate method exists if the daemon is an OSD.
+        Returns:  Pass -> True, Fail -> False
+        """
+        fsid = self.run_ceph_command(cmd="ceph fsid")["fsid"]
+        host = self.fetch_host_node(daemon_type=daemon_type, daemon_id=daemon_id)
+        if daemon_type == "osd":
+            return self.change_osd_state(action=action, target=int(daemon_id))
+        elif daemon_type == "mgr":
+            systemctl_name = (
+                f"ceph-{fsid}@{daemon_type}.{host.hostname}.{daemon_id}.service"
+            )
+        elif daemon_type == "mon":
+            systemctl_name = f"ceph-{fsid}@{daemon_type}.{host.hostname}.service"
+        else:
+            systemctl_name = f"ceph-{fsid}@{daemon_type}.{host.shortname}.service"
+
+        if not host:
+            log.error(f"failed to find host for the {daemon_type} daemon {daemon_id} ")
+            return False
+
+        log.debug(f"Hostname of target host : {host.hostname}")
+
+        # Collecting start time in case of failures
+        init_time, _ = host.exec_command(cmd="sudo date '+%Y-%m-%d %H:%M:%S'")
+        pass_status = True
+        daemon_status, status_desc = self.get_daemon_status(
+            daemon_type=daemon_type, daemon_id=daemon_id
+        )
+
+        if ((daemon_status == 0 or status_desc == "stopped") and action == "stop") or (
+            (daemon_status == 1 or status_desc == "running") and action == "start"
+        ):
+            log.info(
+                f"{daemon_type} daemon {daemon_id} already in desired state: {action}"
+            )
+            return True
+
+        # Executing command to reset the fail count on the host and sleeping for 5 seconds
+        cmd = "systemctl reset-failed"
+        host.exec_command(sudo=True, cmd=cmd)
+        time.sleep(5)
+
+        # Executing command to perform desired action.
+        systemctl_cmd = f"systemctl {action} {systemctl_name}"
+        log.info(
+            f"Performing {action} on {daemon_type}.{daemon_id} on host {host.hostname}. Command {systemctl_cmd}"
         )
-        file_ = host_node.remote_file(sudo=True, file_name=file_name, file_mode="w")
-        file_.write(cont)
-        file_.flush()
-        file_.close()
-
-    # Checking and installing ceph-common package on host
-    try:
-        out, rc = host_node.exec_command(sudo=True, cmd="rpm -qa | grep ceph-common")
-    except Exception:
-        host_node.exec_command(sudo=True, cmd="yum install -y ceph-common")
+        host.exec_command(sudo=True, cmd=systemctl_cmd)
+        time.sleep(5)
+
+        # verifying the daemon state
+        if action in ["start", "stop"]:
```

Comment:
Can we include the  "restart" option in the list?

Reply:
restart is handled in else section.